src/main/java/net/minecraftforge/common/capabilities/CapabilityDispatcher.java

```diff
@@ -5,139 +5,177 @@
+ * the parent object.
+ * 
+ * @param <T> The type of the parent object.
  */
 @ParametersAreNonnullByDefault
 @MethodsReturnNonnullByDefault
-public final class CapabilityDispatcher implements INBTSerializable<CompoundTag>, ICapabilityProvider
+public final class CapabilityDispatcher<T extends ICapabilityProvider> implements INBTSerializable<CompoundTag>, ICapabilityProvider
 {
-    private ICapabilityProvider[] caps;
-    private INBTSerializable<Tag>[] writers;
-    private String[] names;
-    private final List<Runnable> listeners;
 
-    public CapabilityDispatcher(Map<ResourceLocation, ICapabilityProvider> list, List<Runnable> listeners)
+    private final T owner;
+    private final Map<CapabilityType<?>, IAttachedCapabilityProvider<?, T>> caps;
+    private final Map<ResourceLocation, IAttachedCapabilityProvider<?, T>> byName;
+    private boolean isValid = true;
+
+    public CapabilityDispatcher(AttachCapabilitiesEvent<T> event, T owner)
     {
-        this(list, listeners, null);
+        this.caps = event.getCapabilities();
+        this.byName = event.getCapabilitiesByName();
+        this.owner = owner;
     }
 
-    @SuppressWarnings("unchecked")
-    public CapabilityDispatcher(Map<ResourceLocation, ICapabilityProvider> list, List<Runnable> listeners, @Nullable ICapabilityProvider parent)
+    /*********************************
+          ItemStack-Specific Code
+           Warning: Generic Hell
+     *********************************/
+    
+    /**
+     * Internal copy constructor.
+     * @see {@link #copy(ICapabilityProvider)}
+     * @see {@link IItemStackCapabilityProvider#copy(ICapabilityProvider)}
+     */
+    @SuppressWarnings({ "unchecked", "rawtypes" })
+	private CapabilityDispatcher(CapabilityDispatcher<T> other, ItemStack newOwner)
     {
-        List<ICapabilityProvider> lstCaps = Lists.newArrayList();
-        List<INBTSerializable<Tag>> lstWriters = Lists.newArrayList();
-        List<String> lstNames = Lists.newArrayList();
-        this.listeners = listeners;
-
-        if (parent != null) // Parents go first!
+        Map<CapabilityType<?>, IAttachedCapabilityProvider<?, T>> caps = new HashMap<>(other.caps.size(), 1);
+        Map<ResourceLocation, IAttachedCapabilityProvider<?, T>> byName = new HashMap<>(other.byName.size(), 1);
+        for(Map.Entry<ResourceLocation, IAttachedCapabilityProvider<?, T>> entry : other.byName.entrySet())
         {
-            lstCaps.add(parent);
-            if (parent instanceof INBTSerializable)
+            IAttachedCapabilityProvider<?, T> copy = ((IItemStackCapabilityProvider) entry.getValue()).copy(newOwner);
+            // Ideally we would ensure the type and key don't change here, but it's an expensive check that we don't "need" to do.
+            caps.put(copy.getType(), copy);
+            byName.put(copy.getId(), copy);
+        }
+        this.caps = Collections.unmodifiableMap(caps);
```

Comment:
I don't think this is necessary... is there any place where the maps are mutated or exposed?

Reply:
This is for consistency, so that the maps in the fields are always unmodifiable.